### PR TITLE
IO and EXTI driver

### PR DESCRIPTION
**Pull request by ledvinap**
PR for review, please do not merge yet.

Only EXTI handlers are converted, rest of IO is using old code

### CODE REVIEW

**Comment by hydra** [comment]
what needs to happen next for this?

**Comment by ledvinap** [comment]
I'am waiting for your opinion.

I can rebase and squash the commits.

Some refactoring/improving naming conventions may help, I am open to suggestions.

Public interface is important now, the internals should be very easy to change (i tried to hide everything behind opaque types). gcc LTO is quite important to generate reasonable code.

exti is IMO quite usable, even single-wire sonar connection isn't that hard to follow: https://github.com/ledvinap/cleanflight/blob/local/src/main/drivers/sonar_hcsr04.c#L120

**Comment by ledvinap** [comment]
Rebased, will so basic testing tomorrow

**Comment by ledvinap** [comment]
OK, now the PR looks ok ... 
Maybe pull it into separate CF branch and test it?

**Comment by hydra** [comment]
@ledvinap I'll give this a look over soon, thanks for the rebase and new commits.

**Comment by ledvinap** [comment]
OK ... other possible improvements from my repository depend on this,

**Comment by blckmn** [comment]
Just having a look at this @ledvinap and I really like being able to specify the mapping by the devices pin name e.g. PA1...PA13 etc. 

But having a look and it is quite complicated at first glance - and I am not overly keen on magic generation - but that is merely a personal viewpoint. 

Only question is could this be simplified by the function / alternate function (given the functions available on a pin are limited) being mapped/enabled to a particular pin. The relevant function driver (for a given MC) then knows how to configure the function. 

e.g. I2C1_SCL = PB8, I2C1_SDA = PB9.

The I2C1 (device feature) is thus enabled, and the STM32F4 i2c bus driver file then can look up the pin source, clock etc for PB8/PB9 and configure the alternate function accordingly. The I2C1 device feature is then assigned to an alternate (external) device (that will utilise the driver), e.g. BARO_MS5611 (I2C version)

So you would have a list of PINS (PAx, PBx etc), list of MCU device features (I2C1, I2C2, SPI1, SPI2 GPIO, EXTI etc), list of device feature parts (I2C1_SDA, I2C1_SCL, SPI1_MOSI, SPI1_MISO, SPI1_CLK, SPI1_CS1, SPI1_CS2, SPI1,CS3) and a list of external alternate devices (BARO_BMP280, BARO_MS5611, MPU9250, MPU6500 etc).

The mapping is then pin **->** device feature part (-> device feature) **->** external device 

"device feature" is really the grouping of parts (e.g. MISO, MOSI, CLK for SPIs), is not selectable outside the build - as it is specific to device - and would be used in the initialisation to group the pins.

So as an example the MPU9250(SPI) driver file would need on initialisation: the SPI feature (1..5), the chip select for the MPU and the interrupt - irrespective of the SPI pins it is attached to (these are looked up in the bus driver file for the MC used based on the device feature e.g. SPI1).

This approach would require initialisation to cycle through and intialise SPI features, I2C features etc

Not sure if I have described it well. Perhaps I'll have a crack at doing this within raceflight for just i2c to demonstrate.

**Comment by ledvinap** [comment]
@blckmn: 
> But having a look and it is quite complicated at first glance - and I am not overly keen on magic generation - but that is merely a personal viewpoint (and I am a noob so it holds no weight ;)).

I already removed few layers of magic ;-) It can be simplified a bit by removing `DEFIO_REC()` and using `IO_TAG()` everywhere. The cost won't be big, `IO_TAG()` is about one function call (4/8bytes) longer that `IO_TAG()`.

The pin related allocation can be reduced by packing GPIO and PIN into 32bit value (GPIOs are always aligned, (up to 12) low bits can be used to identify bit). It will be probably even faster, replacing memory reads with mask/shits operations. But this change is internal and hidden from user. Per-pin allocated memory will remain to handle resource management.

If you consider public interface only, it is too complicated?

Ad alternate functions:
It may be little bit more complicated on F3, each pin has 16 possible alternate functions (but not all positions are used).
Also it should be possible to configure some peripherals IO mapping at runtime. It is useful to remap USART to alternate pins, possibly when changing device owner (USART1 - MSP when disarmed, BLACKBOX when armed, use alternate pins (better accessibility on NAZE) for BLACKBOX). 

One possibility is to add list of possible pin mapping + alternate function/remap for each device driver _Def structure. Configuration may be selected by pin or by index.
Or limit runtime-configurable choices (few reasonable variants) and specify them in target.h (possibly indirectly, using CPU/PACKAGE specific macros).

Not sure about this now .. it would be nice to allow external peripherals, but then most general case should be assumed ..

Can you write some pseudocode for your system, please?

**Comment by blckmn** [comment]
OK @ledvinap I am on board. See https://github.com/rs2k/raceflight/pull/79

If I could work out how to have brought your commits over I would have - apologies. Hopefully the commit comments provide adequate attribution for the work. I've implemented it for LEDs (on the F4) and also started work on it for i2c configuration. Will look to implement Beeper etc shortly.

**Comment by ledvinap** [comment]
@blckmn : Glad to see someone using it.
Most of CF is converted to new IO interface in my local branch (https://github.com/ledvinap/cleanflight).
Timer HAL is almost finished too, I'll prepare PR when/if IO is merged and tested ..

**Comment by ledvinap** [comment]
@blckmn : I use magit, applying commits from other branch is trivial there.

**Comment by hydra** [comment]
awesome to see more work on this guys!

**Comment by ledvinap** [comment]
Should I prepare this code for merge?

**Comment by blckmn** [comment]
I am happy with it - though I don't think you need the auto generation - its a once off.

**Comment by ledvinap** [comment]
@blckmn : Sorry, I am not sure what does 'once off' mean in this context.

If you mean that generation is only one-time event - what if generated source needs some change? With generator it's easy to change the script. Without it lot of places are affected, with hude changeset and possible errors ..
Generated files are in git so there is no need to generate files for common users ..

**Comment by blckmn** [comment]
Fair point if you feel it will get changed over time.

**Comment by ledvinap** [comment]
My idea is interface that completely hides internal implementation.

**Comment by ledvinap** [comment]
Is it ok to force-push into this PR? Or should I create new branch with rebased commits?

**Comment by hydra** [comment]
ok #1984 was merged, closing this.
